src/services/refactors/convertObjectDestruction.ts

```diff
@@ -0,0 +1,378 @@
+/* @internal */
+namespace ts.refactor {
+    const refactorName = "Introduce destruction";
+    const actionNameIntroduceObjectDestruction = "Convert access to destruction";
+    registerRefactor(refactorName, { getAvailableActions, getEditsForAction });
+
+    function getAvailableActions(context: RefactorContext): readonly ApplicableRefactorInfo[] {
+        const { file, program, cancellationToken } = context;
+        const isJSFile = isSourceFileJS(file);
+        if (isJSFile || !cancellationToken) return emptyArray;
+
+        const info = getInfo(context, file, program.getTypeChecker(), program, cancellationToken, /*resolveUniqueName*/ false);
+        if (!info) return emptyArray;
+
+        return [{
+            name: refactorName,
+            description: actionNameIntroduceObjectDestruction,
+            actions: [{
+                name: refactorName,
+                description: actionNameIntroduceObjectDestruction
+            }]
+        }];
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        Debug.assert(actionName === refactorName);
+        const { file, program, cancellationToken } = context;
+        const isJSFile = isSourceFileJS(file);
+
+        const emptyResult: RefactorEditInfo = { edits: [] };
+        if (isJSFile || !cancellationToken) return emptyResult;
+
+        const info = getInfo(context, file, program.getTypeChecker(), program, cancellationToken, /*resolveUniqueName*/ true);
+        if (!info) return emptyResult;
+
+        const edits = textChanges.ChangeTracker.with(context, t => doChange(t, file, info));
+        return { renameFilename: undefined, renameLocation: undefined, edits };
+    }
+
+    function getUniqueNumericAccessVariable(name: string | number, file: SourceFile) {
+        const tempName = `index_${name}`;
+        return isFileLevelUniqueName(file, tempName) ? tempName : getUniqueName(tempName, file);
+    }
+
+    /**
+     * `Dense` means we use array literal pattern to destruction the expression.
+     * We allowed index up to 15 to avoid many omit expression.

```

Comment:
Is this arbitrary? I would be more conservative and additionally not offer the destructure if the number of omitted indices is greater than the included indices. `const [, , , index_3, , , , , index_8, , , , , , index_14] = item;` looks very weird to me.

Reply:
How about not offer refactor if not trigger by 'invoke'. 
Otherwise convert to object destruction.